finesse/hardware/st10_controller.py

```diff
@@ -0,0 +1,305 @@
+        """
+        step = self._request_value("SP")
+        try:
+            return int(step)
+        except ValueError:
+            raise ST10ControllerError(f"Invalid value for step received: {step}")
+
+    @step.setter
+    def step(self, step: int) -> None:
+        """Move the stepper motor to the specified absolute position.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        self._write_check(f"FP{step}")
+
+    def _send_string(self, string: str) -> None:
+        """Request that the device sends string when operations have completed."""
+        self._write_check(f"SS{string}")
+
+    def _read(self) -> str:
+        """Read the next message from the device.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        raw = self.serial.read_until(b"\r")
+
+        # Check that it hasn't timed out
+        if not raw:
+            raise SerialTimeoutException()
+
+        logging.debug(f"(ST10) <<< {repr(raw)}")
+
+        try:
+            return raw[:-1].decode("ascii")
+        except UnicodeDecodeError:
+            raise ST10ControllerError(f"Invalid message received: {repr(raw)}")
+
+    def _write(self, message: str) -> None:
+        """Send the specified message to the device.
+
+        Raises:
+            SerialException: Error communicating with device
+            UnicodeEncodeError: Malformed message
+        """
+        data = f"{message}\r".encode("ascii")
+        logging.debug(f"(ST10) >>> {repr(data)}")
+        self.serial.write(data)
+
+    def _write_check(self, message: str) -> None:
+        """Send the specified message and check whether the device returns an error.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+            UnicodeEncodeError: Message to be sent is malformed
+        """
+        self._write(message)
+        self._check_response()
+
+    def _check_response(self) -> None:
+        """Check whether the device has returned an error.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        response = self._read()
+
+        # These values are referred to as "ack" and "qack" in the old program (nack is
+        # "?"). I don't know what qack is. Could it mean clockwise/anticlockwise
+        # movement?
```

Comment:
I think this clarification is actually a bit confusing. We are not using any of those terms here, so why to include an explanation about them - related to a program no one (but you) have the source code for.

Reply:
Maybe it's not the best explanation. But I feel like ack and nack are fairly standard terms, so it's obvious what the meaning was, but it is a mystery as to why there are two versions of "ack". This seems to be out of scope for the manual I linked to, so probably the only person who knows is Alan, who built the hardware and the old software and is now retired (I think).

I think it is worth flagging up in the comment that there are two "ack" responses and that they may mean slightly different things, even if we currently treat them the same in the program. We don't have any documentation for this so the closest thing is the source code for the old program.

Reply:
What do you think?

Reply:
Well, "it's obvious what the meaning was" might be an overstatement. I don't know what the meaning of those terms is, so I would suggest you elaborate on that.

Reply:
Fair enough, sorry. I'll write a better explanation.

Reply:
Ha, it turns out that the manual does have an explanation for what these responses mean -- I just wasn't looking hard enough.